### PR TITLE
Fix regression on SwiftPackageManager packages defining file resources with copy rule

### PR DESCRIPTION
**Pull request by alexanderwe**
Resolves: https://github.com/tuist/tuist/issues/4796 and regression mentioned in https://github.com/tuist/tuist/pull/4733

### Short description 📝

This PR tries to fix a regression on SPM projects where resources are included with the `.copy` rule. It seems that the previous implementation failed for the following "type" of resources:

* Opaque directories that should be handled like files (e.g. xcdatamodel)
* Single files which are not directories 

In order to not break unit tests it is now required for SPM generation related unit tests that files and directories declared as resources need to be really present on the file system during the test. Otherwise `AbsolutePath.isDirectory` and `AbsolutePath.isFile` will not work correctly and report fails positive. 

I also changed the `app_with_spm_dependencies` fixture project to include the two cases mentioned above. Unfortunately they cannot be checked at build time. Some need to run the project to verify the files are accessible. Or to take a look at the `Copy Bundle resources` manually. 

### How to test the changes locally 🧐

Try to run `tuist fetch` and `tuist generate` for the following repositories

* https://github.com/salqueng/tuist-with-stream-chat-swift-bug (opaque files copy case)
* https://github.com/Arafo/tuist-spm-resources-copy-issue (single file copy case)
* https://github.com/alexanderwe/tuist-spm-resources-copy-issue (directory copy case)
* Generation and build of these projects should not fail

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] The title of the PR will be used as changelog entry, please make sure it is clear and suitable
- [ ] In case the PR introduces changes that affect users, the documentation has been updated
- [ ] Contributors have checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`

### CODE REVIEW

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift [hunk 1]
```diff
@@ -698,6 +698,10 @@ extension ResourceFileElements {
 
             switch $0.rule {
             case .copy:
+                // Single files or opaque directories are handled like a .process rule
+                if resourceAbsolutePath.isFile || resourceAbsolutePath.isOpaqueDirectory {
+                    return handleProcessResource(resourceAbsolutePath: resourceAbsolutePath)
+                }
                 return handleCopyResource(resourceAbsolutePath: resourceAbsolutePath)
```
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:705]:
```suggestion
                } else {
                  return handleCopyResource(resourceAbsolutePath: resourceAbsolutePath)
                }
```

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift [hunk 2]
```diff
@@ -92,6 +92,28 @@ extension AbsolutePath {
...
+
+    /// This property returns true when the path points to a directory.
+    public var isDirectory: Bool {
+        (try? self.url.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory == true
+    }
+
+    /// This property returns true when the path points to a file.
+    public var isFile: Bool {
+        return !isDirectory
+    }
```
Comment (danyf90) [inline Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift:115]: These are already defined in FileHandler.isFolder

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift [hunk 3]
```diff
@@ -92,6 +92,28 @@ extension AbsolutePath {
...
+    /// An opaque directory is a directory that should be treated like a file, therefor ignoring its content.
+    /// I.e.: .xcassets, .xcdatamodeld, etc...
+    /// This property returns true when a file is such a directory.
+    public var isOpaqueDirectory: Bool {
+        if let `extension` = self.extension,
+           Self.opaqueDirectoriesExtensions.contains(`extension`)
+        {
+            return true
+        }
+        return false
```
Comment (danyf90) [inline Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift:104]:
```suggestion
        return Self.opaqueDirectoriesExtensions.contains(self.extension ?? "")
```

**Review by danyf90 — APPROVED** [review]

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift (hunk 1 above)
Comment (alexanderwe) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: Fixed

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift (hunk 2 above)
Comment (alexanderwe) [reply Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift]: Using `FileHandler.isFolder` now

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift (hunk 3 above)
Comment (alexanderwe) [reply Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift]: Fixed and listed

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift [hunk 4]
```diff
@@ -82,16 +82,19 @@ extension AbsolutePath {
...
             currentDirectory = currentDirectory.parentDirectory
         }
         return false
     }
 
+    /// An opaque directory is a directory that should be treated like a file, therefor ignoring its content.
+    /// I.e.: .xcassets, .xcdatamodeld, etc...
+    /// This property returns true when a file is such a directory.
+    public var isOpaqueDirectory: Bool {
+        Self.opaqueDirectoriesExtensions.contains(self.extension ?? "")
```
Comment (danyf90) [inline Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift:95]:
```suggestion
        [
          "xcassets",
          "scnassets",
          "xcdatamodeld",
          "docc",
          "playground",
          "bundle",
          "mlmodelc",
      ]
      .contains(self.extension ?? "")
```

**Review by danyf90 — APPROVED** [review]

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift (hunk 4 above)
Comment (alexanderwe) [reply Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift:95]: Good point, removed `opaqueDirectoriesExtensions` and integrated it into `isOpaqueDirectory`

**Pull request merged**